### PR TITLE
feat: browserContext.on('console')

### PR DESCRIPTION
**Pull request by dgozman**

### CODE REVIEW

File: docs/src/api/class-consolemessage.md [hunk 1]
```diff
@@ -125,6 +125,12 @@ List of arguments passed to a `console` function call. See also [`event: Page.co
 
 URL of the resource followed by 0-based line and column numbers in the resource formatted as `URL:line:column`.
 
+## method: ConsoleMessage.page
+* since: v1.33
+- returns: <[Page]|[null]>
```
Comment (mxschmitt) [inline docs/src/api/class-consolemessage.md:130]: how can Page be null if its not nullable on the server side?

Comment (dgozman) [reply docs/src/api/class-consolemessage.md:130]: For future things like ServiceWorker.

File: packages/playwright-core/src/server/frames.ts [hunk 2]
```diff
@@ -1028,8 +1028,8 @@ export class Frame extends SdkObject {
     let cspMessage: ConsoleMessage | undefined;
     const actionPromise = func().then(r => result = r).catch(e => error = e);
     const errorPromise = new Promise<void>(resolve => {
-      listeners.push(eventsHelper.addEventListener(this._page, Page.Events.Console, (message: ConsoleMessage) => {
-        if (message.type() === 'error' && message.text().includes('Content Security Policy')) {
+      listeners.push(eventsHelper.addEventListener(this._page._browserContext, BrowserContext.Events.Console, (message: ConsoleMessage) => {
```
Comment (pavelfeldman) [inline packages/playwright-core/src/server/frames.ts:1031]: Can we fix this as a drive-by? Looks like a layering violation.

File: packages/playwright-core/src/server/page.ts [hunk 3]
```diff
@@ -208,12 +208,18 @@ export class Page extends SdkObject {
     }
     this._initialized = true;
     this.emitOnContext(contextEvent, this);
-    // I may happen that page initialization finishes after Close event has already been sent,
+
+    for (const message of this._consoleMessagesBeforeInitialized)
```
Comment (pavelfeldman) [inline packages/playwright-core/src/server/page.ts:212]: Comment would be nice

File: packages/playwright-core/src/server/page.ts [hunk 4]
```diff
@@ -351,10 +357,15 @@ export class Page extends SdkObject {
...
     const message = new ConsoleMessage(this, type, text, args, location);
     const intercepted = this._frameManager.interceptConsoleMessage(message);
-    if (intercepted || !this.listenerCount(Page.Events.Console))
+    if (intercepted) {
       args.forEach(arg => arg.dispose());
+      return;
+    }
+
+    if (this._initialized)
+      this._browserContext.emit(BrowserContext.Events.Console, message);
```
Comment (pavelfeldman) [inline packages/playwright-core/src/server/page.ts:366]: this.emitOnContext

File: packages/playwright-core/src/server/page.ts [hunk 5]
```diff
@@ -141,6 +140,7 @@ export class Page extends SdkObject {
   private _closedPromise = new ManualPromise<void>();
   private _disconnected = false;
   private _initialized = false;
+  private _consoleMessagesBeforeInitialized: ConsoleMessage[] = [];
```
Comment (pavelfeldman) [inline packages/playwright-core/src/server/page.ts:143]: We are going to have a lot of those, create a vector of typed evens?
